File: 08/08b.py
class InstructionIndexOutOfBoundsError(Exception):
    pass

def floyd(f, x0):
    tortoise = f(x0)
    hare = f(f(x0))
    while tortoise[0] != hare[0]:
        tortoise = f(tortoise)
        hare = f(f(hare))
  
    mu = 0
    tortoise = x0
    while tortoise[0] != hare[0]:
        tortoise = f(tortoise)
        hare = f(hare)
        mu += 1
 
    lam = 1
    hare = f(tortoise)
    while tortoise[0] != hare[0]:
        hare = f(hare)
        lam += 1
 
    return lam, mu

def parse_instruction(line):
    line = line.strip()
    [instruction, value] = line.split(' ')
    return instruction, int(value)

def main():
    with open('input.txt') as f:
        instructions = [parse_instruction(line) for line in f.readlines()]

        def execute_instruction(args):
            index, acc = args

            if index >= len(instructions):
                raise InstructionIndexOutOfBoundsError()

            instruction, value = instructions[index]
            if instruction == 'nop':
                return index + 1, acc
            elif instruction == 'acc':
                return index + 1, acc + value
            elif instruction == 'jmp':
                return index + value, acc
            else:
                raise Exception('Unknown instruction code')

        for count, (instruction, value) in enumerate(instructions):
            if instruction == 'acc':
                continue
            
            new_instruction = 'jmp' if instruction == 'nop' else 'nop'
            instructions[count] = (new_instruction, value)

            try:
                lam, mu = floyd(execute_instruction, (0, 0))

                # restore initial instruction
                instructions[count] = (instruction, value)
            except InstructionIndexOutOfBoundsError:
                break

        index = 0
        acc = 0
        while True:
            try:
                index, acc = execute_instruction((index, acc))
            except InstructionIndexOutOfBoundsError:
                print(acc)
                break

if __name__ == '__main__':
    main()

File: 08/08b.hs
import Data.Map (Map)
import qualified Data.Map as Map

main = do
  content <- readFile "input.txt"

  let linesOfInput = lines content
  let instructions = map parseInstruction linesOfInput
  let acc = findNonCyclical instructions 0
  print acc

parseInstruction :: String -> (String, Int)
parseInstruction line = (instruction, value)
  where lineWords = words line
        instruction = lineWords !! 0
        trimmedValue = dropWhile (\c -> c == '+') (lineWords !! 1)
        value = read trimmedValue :: Int

findNonCyclical :: [(String, Int)] -> Int -> Int
findNonCyclical instructions index
  | hasCycle == True = findNonCyclical instructions (index + 1)
  | otherwise = acc
  where (instruction, value) = instructions !! index
        swappedInstructions = swapInstructions instructions index
        (hasCycle, acc) = isCyclical swappedInstructions Map.empty 0 0

isCyclical :: [(String, Int)] -> Map Int Bool -> Int-> Int -> (Bool, Int)
isCyclical instructions visited index acc
  | index >= (length instructions) = (False, acc)
  | (Map.lookup index visited) == Nothing = isCyclical instructions newVisited newIndex newAcc
  | otherwise = (True, acc)
  where (newIndex, newAcc) = executeInstruction instructions index acc
        newVisited = Map.insert index True visited

executeInstruction :: [(String, Int)] -> Int -> Int -> (Int, Int)
executeInstruction instructions index acc
  | instruction == "nop" = (index + 1, acc)
  | instruction == "acc" = (index + 1, acc + value)
  | instruction == "jmp" = (index + value, acc)
  where (instruction, value) = instructions !! index

swapInstructions :: [(String, Int)] -> Int -> [(String, Int)]
swapInstructions instructions index
  | instruction == "acc" = instructions
  | instruction == "nop" = replaceElement instructions index ("jmp", value)
  | instruction == "jmp" = replaceElement instructions index ("nop", value)
  where (instruction, value) = instructions !! index

replaceElement :: [(String, Int)] -> Int -> (String, Int) -> [(String, Int)]
replaceElement instructions index newElement = take index instructions ++ newElement : drop (index + 1) instructions
